### ISSUE TITLE
Unexpected value 'Property' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.SymbolKind'

### ISSUE THREAD

**Issue by Symonz** [Microsoft/AL#3435]
- AL Extension Version: 0.12.17091
- Server Version: W1 11.0 (22292)

I just finished converting a big verticalization to extension 2.0 on a world wide environment, passing from 2000 errors to 0 errors. Just 35 warnings remain (all about the Product Group table marked for removal).

Unfortunately, the creation of the extension fails with the following low level error, that doesn't give enough information about the reason.

Could you please provide at least a clue on the cause of this error? Thanks.
______________________________________________________________________________________________________
```
Processing of message 'al/createPackage' failed with error: 'Unexpected value 'Property' of type 'Microsoft.Dynamics.Nav.CodeAnalysis.SymbolKind''
StackTrace:
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindNonMethod(NameSyntax node, Symbol symbol, Boolean isError, BoundExpression receiver) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Expressions.cs:line 1349
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindIdentifier(IdentifierNameSyntax node, Boolean invoked, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Expressions.cs:line 1010
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindArgumentAndName(AnalyzedArguments result, Boolean hadError, ExpressionSyntax argumentExpression) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Expressions.cs:line 1144
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindArgumentsAndNames(SeparatedSyntaxList`1 arguments, AnalyzedArguments result) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Expressions.cs:line 1124
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindInvocationExpression(InvocationExpressionSyntax node, DiagnosticBag diagnostics, Boolean isStatement) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Invocation.cs:line 31
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindExpressionStatement(SyntaxNode node, ExpressionSyntax syntax, Boolean allowsAnyExpression, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 517
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 83
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindStatements(SyntaxList`1 syntaxStatements, DiagnosticBag diagnostics, Binder binder) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 53
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics, Binder blockBinder) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 33
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 80
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindIfStatement(IfStatementSyntax node, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 542
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 92
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindStatements(SyntaxList`1 syntaxStatements, DiagnosticBag diagnostics, Binder binder) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 53
   at Microsoft.Dynamics.Nav.CodeAnalysis.Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics, Binder blockBinder) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Binder\Binder_Statements.cs:line 33
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.BindMethodBody(MethodSymbol method, DiagnosticBag diagnostics) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 463
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMethod(MethodSymbol methodSymbol) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 385
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMembers(ContainerSymbol containingType) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 309
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileObject(ObjectTypeSymbol containingType) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 235
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.<>c__DisplayClass17_0.<CompileObjectAsTask>b__0() in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 216
   at Microsoft.Dynamics.Nav.CodeAnalysis.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Utilities\UICultureUtilities.cs:line 152
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.WaitForWorkers() in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 161
   at Microsoft.Dynamics.Nav.CodeAnalysis.MethodCompiler.CompileMethodBodies(Compilation compilation, ModuleBuilder moduleBeingBuiltOpt, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compiler\MethodCompiler.cs:line 145
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.CompileImpl(ModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 810
   at Microsoft.Dynamics.Nav.CodeAnalysis.Compilation.Compile(ModuleBuilder moduleBuilder, Boolean emittingPdb, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\Compilation\Compilation.cs:line 787
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.PackageTheModule(Compilation compilation, CompilerTextWriter consoleOutput, CancellationToken cancellationToken, ErrorLogger errorLogger, AnalyzerDriver analyzerDriver) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 531
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.RunCore(CompilerTextWriter consoleOutput, ErrorLogger errorLogger, CancellationToken cancellationToken) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 413
   at Microsoft.Dynamics.Nav.CodeAnalysis.CommandLine.CommonCompiler.Run(CompilerTextWriter consoleOutput, CancellationToken cancellationToken) in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.CodeAnalysis\CommandLine\CommonCompiler.cs:line 312
   at Microsoft.Dynamics.Nav.EditorServices.Protocol.LanguageServer.Extensions.CreatePackageRequestHandler.<HandleAsync>d__3.MoveNext() in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Protocol\LanguageServer\Extensions\CreatePackageRequestHandler.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Dynamics.Nav.EditorServices.Protocol.RequestRegistry.<Process>d__6.MoveNext() in C:\a\36\s\source\Prod\Microsoft.Dynamics.Nav.EditorServices.Protocol\Endpoints\RequestRegistry.cs:line 59
```

**Comment by thpeder** [comment]
Hi @Symonz,
Simply by looking at the stack trace It seems like you have a procedure with an if statement and in that there is a procedure invocation where it finds a property for one of the arguments to it.
I hope that helps.

**Comment by Symonz** [comment]
Hi Thomas,
Thank you very much for your hint. We have approximately 16.700 "if" statements, so it would be appreciated that the error would report the file and file line (or at least the code statement) that provokes the exception.
By the way, we are trying to find the problematic if statement(s) that raise the above error. Could you please be more precise on the last part of your answer (

> where it finds a property for one of the arguments to it

)?
An example of such a statement would be of help.
Thank you in advance.
Simon

**Comment by thpeder** [comment]
It is kinda hard to follow the stack trace and know for sure what how the data looks like. 
But I guess you might have a variable named the same as a property.

Could you try getting the latest extension from the marketplace and check if it reproduces there.
Thanks.

**Comment by Symonz** [comment]
Hi Thomas,

I've installed the latest version of the extension, but the error still remains. I'm trying searching the code for all possibly problematic if statements via regular expressions. Anyway, I cannot get rid of that error, considering the relevant number of if statements (about 16.000).

I've tried to simulate that error in order to be sure of the scenario that I need to search. It seems that the problem raises not just when a variable with the same name of a property is used in an if statement, but when that variable is not even defined.

The syntax checker doesn't show any error, but the compiler raises the error I posted.

I think the syntax checker must show this kind of error at the beginning, or the following error should show at least the object that provokes it.

To refine my search, is it possible to get a list of AL Object properties?

Do you think to release a patch that show this kind of error in the VS Editor, before the actual compilation?

Thanks in advance.
Simon

**Comment by thpeder** [comment]
Hi @Symonz,
If you are able to write a small example that reproduces the problem, then could you please share that so we can get the problem fixed, and have the compiler simply give an error on the at the location of the problem instead of crashing.
We are discussion internally how we can improve on these kind of situations where you want to know more about what the compiler does in order to pinpoint the source of the problem, but we don't have any specific feature in the pipeline unfortunately.

Thanks.

**Comment by thpeder** [comment]
Might be related to #3926

**Comment by StanislawStempin** [comment]
Closing due to inactivity and since the original reported issue is on NAV 2018 and not the latest Dev Preview of Business Central.
